sdk/eventhub/event-hubs/src/eventProcessor.ts

```diff
@@ -400,73 +413,34 @@ export class EventProcessor {
-
-        for (const ownership of partitionOwnership) {
-          if (isAbandoned(ownership)) {
-            abandonedMap.set(ownership.partitionId, ownership);
-            continue;
-          }
-
-          partitionOwnershipMap.set(ownership.partitionId, ownership);
-        }
         const { partitionIds } = await this._context.managementSession!.getEventHubProperties({
           abortSignal: abortSignal
         });
 
-        if (abortSignal.aborted) {
-          return;
-        }
-
-        if (partitionIds.length > 0) {
-          const partitionsToClaim = loadBalancer.loadBalance(
-            this._id,
-            partitionOwnershipMap,
-            partitionIds
-          );
-          if (partitionsToClaim) {
-            for (const partitionToClaim of partitionsToClaim) {
-              let ownershipRequest: PartitionOwnership;
-
-              if (abandonedMap.has(partitionToClaim)) {
-                ownershipRequest = this._createPartitionOwnershipRequest(
-                  abandonedMap,
-                  partitionToClaim
-                );
-              } else {
-                ownershipRequest = this._createPartitionOwnershipRequest(
-                  partitionOwnershipMap,
-                  partitionToClaim
-                );
-              }
-
-              await this._claimOwnership(ownershipRequest, abortSignal);
-            }
-          }
-        }
+        // Renew the EventProcessor's partition ownerships that are still active.
+        // await this._renewPartitionOwnership(abortSignal);
+
+        await this._performLoadBalancing(loadBalancingStrategy, partitionIds, abortSignal);
       } catch (err) {
-        logger.warning(`[${this._id}] An error occured within the EventProcessor loop: ${err}`);
+        logger.warning(
+          `[${this._id}] An error occurred within the EventProcessor loop: ${err?.message}`
+        );
         logErrorStackTrace(err);
         // Protect against the scenario where the user awaits on subscription.close() from inside processError.
         await Promise.race([this._handleSubscriptionError(err), cancelLoopPromise]);
       } finally {
-        // sleep for some time, then continue the loop again.
+        // Sleep for some time, then continue the loop.
+        const iterationDeltaInMs = Date.now() - iterationStartTimeInMs;
+        const delayDurationInMs = Math.max(this._loopIntervalInMs - iterationDeltaInMs, 0);
         logger.verbose(
-          `[${this._id}] Pausing the EventProcessor loop for ${this._loopIntervalInMs} ms.`
+          `[${this._id}] Pausing the EventProcessor loop for ${delayDurationInMs} ms.`
         );
-        // swallow the error since it's fine to exit early from delay
-        await delayWithoutThrow(this._loopIntervalInMs, abortSignal);
+        // Swallow the error since it's fine to exit early from the delay.
+        await delayWithoutThrow(delayDurationInMs, abortSignal);
```

Comment:
I'm not sure I understand what the concern you're getting at is.

In general this delay is in place just so that we have a gap between each iteration of our load balancing loop. If the duration happens to be 0 (because this iteration of the loop happened to exceed the loopIntervalInMs) it's still ok to yield, but we don't strictly speaking have to.

Reply:
Oh sorry I didn't explain this very well.

My first thought was "can we skip the delay when it's 0?" but then I thought that not yielding might not let background tasks complete, so it's useful.

The other remark is that in the browser sometimes setTimeout(0) acts like setTimeout(16) due to various implementation details.

Reply:
We do want to make sure that the loop yields at some point since there will be background tasks running.

Yeah, 16 ms isn't a big deal here. I'd expect that most of the time, the `delayDurationInMs` will be larger than 0 (the default delay is 10 seconds, and so we're subtracting the time it takes to make 1 call to event hubs and then some calls potentially to a service like storage if the user is using storage to checkpoint, which I wouldn't expect to take 10 seconds.) We're ultimately trying to limit how many service calls we make while still doing it often enough that we're able to detect when we need to make changes to reach a balanced state, so if it takes just a little longer that's fine.

I like having this yield even if delay is 0 because then we don't have to do extra checks just to handle the '0' case, and because then we're guaranteed at least 1 yield for background tasks :smile: